README.md
# Lights Out
[Lights Out](https://en.wikipedia.org/wiki/Lights_Out_(game)) is a simple game
where you have a grid of lights and you try to turn them all off. Clicking a
light toggles the light itself and all adjacent lights (not including diagonally
adjacent). This is an implementation using Clojure and
[Seesaw](https://en.wikipedia.org/wiki/Lights_Out_(game)).

## Installation

Just clone this repo and make sure you have [Leinengen](http://leiningen.org/).

## Usage

Run the command `lein run -m lights-out.core`. The game will start and you can
click lights to your heart's content.

## Options

None yet.

## Examples

![An image of the board](screenshots/board.png)

### Bugs

Currently, all you can do is click the lights. You won't be notified if you win
and there's no option to restart or start a new game.

### Any Other Sections
### That You Think
### Might be Useful

## License

Copyright © 2016 Julius Alexander IV

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased][unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2016-03-27
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2016-03-27
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[unreleased]: https://github.com/your-name/lights-out/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/lights-out/compare/0.1.0...0.1.1


test/lights_out/core_test.clj
(ns lights-out.core-test
  (:require [clojure.test :refer :all]
            [lights-out.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


src/lights_out/stable.clj
(ns lights-out.core
  (:gen-class))
(use 'seesaw.core)

(native!)

; Create the frame that everything will be drawn on.
(def f (frame :title "Lights Out"))
; Define the number of lights in the board as a constant.
(def ^:const board-size 5)
(def ^:const total-tiles (* board-size board-size))

(declare update-board)
(declare draw-buttons)
(declare make-button)

; Create a blank board with random tiles turned on.
(def board
  (into [] (map (fn [e] (= (rand-int 2) 1)) (range total-tiles))))

; Create a list of buttons from the board data structure.
(defn draw-buttons [bs]
  (map-indexed #(make-button %1 (if %2 "#00A1FF" "#666666")) bs))

; Create a button.
(defn make-button [i color]
  (button :listen [:action (fn [e] (update-board i))]
          :preferred-size [40 :by 40]
          :background color))

(def button-widget (grid-panel :items (draw-buttons board) :columns 5))
(defn update-button-widget! [bs]
  (def button-widget (grid-panel :items (draw-buttons bs) :columns 5)))

; Calculate the indeices of the affected tiles.
(defn get-updateable [i bsize ttiles]
  (filter #(and (> % 0) (< % ttiles))
          [i (+ i 1) (- i 1) (+ i bsize) (- i bsize)]))

; Toggle the light in board |b| at position |i|.
(defn toggle-cell [i b]
  (def board (assoc board i (not (get board i)))))

(defn print-board []
  (doseq [i board]
    (println i)))

; Update the board when button |i| is clicked.
(defn update-board [i]
  (for [t (get-updateable i board-size total-tiles)]
    (toggle-cell t board))
  (print-board)
  (update-button-widget! board)
  (repaint! button-widget))


; Load the board into the frame.
(defn display [content]
  (config! f :content content)
  content)

; Entry point.
(defn -main
  "Show the screen"
  [& args]
  (update-button-widget! board)
  (display button-widget)
  (-> f pack! show!))


src/lights_out/core.clj
(ns lights-out.core
  (:gen-class)
  (:use [seesaw.core :as seesaw])
  (:use [lights-out.util :as util]))

(native!)

; Create the frame that everything will be drawn on.
(def game-frame (frame :title "Lights Out"))
; Define the number of lights in the board as a constant.
(def ^:const board-size 5)
(def ^:const total-tiles (* board-size board-size))

(declare update-board)
(declare make-light-list)
(declare make-light)

; Create a blank board with random tiles turned on.
(def board-data
  (to-array-2d
    (map (fn [_] (util/rand-bool-coll board-size)) (range board-size))))

(defn make-light-list
  "Create a list of buttons from the board data structure."
  [vals]
  (apply concat
    (util/amap-indexed
      #(make-light %1 %2 (if %3 "#00A1FF" "#666666"))
      vals)))

(defn make-light
  "Create a button."
  [row col color]
  (seesaw/button :listen [:action (fn [_] (update-board row col))]
                 :preferred-size [40 :by 40]
                 :background color))

(defn new-game-window
  "Create a grid of lights based on the values in |board-vals|.
  Note: |board-vals| is a 2d array and must be flattened before it can be used
  with a seesaw function."
  [board-vals]
  (seesaw/grid-panel :items (make-light-list board-vals)
                     :columns 5))

(defn get-updateable
  "Calculate the indices of the affected tiles when the light at |row|, |col| is
  clicked."
  [row col bsize]
  (filter (fn [v] (every? #(<= 0 % (- bsize 1)) v))
          [[row col]
           [(- row 1) col]
           [(+ row 1) col]
           [row (- col 1)]
           [row (+ col 1)]]))

(defn toggle-cell
  "Toggle the light in board |b| at position |i|."
  [row col]
  (let [opp (not (aget board-data row col))]
    (aset board-data row col opp)))

(defn update-board
  "Update the board when button |i| is clicked."
  [row col]
  (doseq [coord (get-updateable row col board-size)]
    (apply toggle-cell coord))
  (seesaw/config! game-frame :content (new-game-window board-data)))

(defn display
  "Load the board into the frame."
  [content]
  (seesaw/config! game-frame :content content)
  content)

; Entry point.
(defn -main
  "Show the screen"
  [& args]
  (display (new-game-window board-data))
  (-> game-frame pack! show!))
